IncludeToolbox/IncludeFormatter/IncludeFormatter.cs

```diff
@@ -86,26 +87,85 @@ public static void FormatSlashes(IncludeLineInfo[] lines, FormatterOptionsPage.S
             }
         }
 
-        public static void SortIncludes(IncludeLineInfo[] lines, FormatterOptionsPage.TypeSorting typeSorting, bool regexIncludeDelimiter, string[] precedenceRegexes, string documentName)
+        public static void SortIncludes(IncludeLineInfo[] lines, FormatterOptionsPage settings, string documentName)
         {
-            var comparer = new IncludeComparer(precedenceRegexes, documentName);
-            var sortedIncludes = lines.Where(x => x.LineType != IncludeLineInfo.Type.NoInclude).OrderBy(x => x.IncludeContentForRegex(regexIncludeDelimiter), comparer);
+            FormatterOptionsPage.TypeSorting typeSorting = settings.SortByType;
+            bool regexIncludeDelimiter = settings.RegexIncludeDelimiter;
+            bool blankAfterRegexGroupMatch = settings.BlankAfterRegexGroupMatch;
+            bool removeEmptyLines = settings.RemoveEmptyLines;
 
-            if (typeSorting == FormatterOptionsPage.TypeSorting.AngleBracketsFirst)
-                sortedIncludes = sortedIncludes.OrderBy(x => x.LineType == IncludeLineInfo.Type.AngleBrackets ? 0 : 1);
-            else if (typeSorting == FormatterOptionsPage.TypeSorting.QuotedFirst)
-                sortedIncludes = sortedIncludes.OrderBy(x => x.LineType == IncludeLineInfo.Type.Quotes ? 0 : 1);
+            var comparer = new IncludeComparer(settings.PrecedenceRegexes, documentName);
+            string[] precedenceRegexes = comparer.PrecedenceRegexes;
+
+            var sortedIncludes = lines.Where(x => x.LineType != IncludeLineInfo.Type.NoInclude).OrderBy(x => x.IncludeContentForRegex(regexIncludeDelimiter), comparer).ToArray();
+
+            var mapSortedIndexToRealIndex = new Dictionary<int, int>();
+            {
+                int sortedIdx = 0;
+                for (int realIdx = 0; realIdx < lines.Length && sortedIdx < sortedIncludes.Length; ++realIdx)
+                {
+                    if (lines[realIdx].LineType != IncludeLineInfo.Type.NoInclude)
+                    {
+                        mapSortedIndexToRealIndex.Add(sortedIdx, realIdx);
+                        ++sortedIdx;
+                    }
+                }
+            }
 
-            int incIdx = 0;
-            var sortedIncludesArray = sortedIncludes.ToArray();
-            for (int allIdx = 0; allIdx < lines.Length && incIdx < sortedIncludesArray.Length; ++allIdx)
+            // Optionally insert newlines between regex match groups
```

Comment:
This block does the work to add blank lines. It more or less just groups the includes by which regex they match and prepends a newline on the first of each group (except for the first group).